src/Dapr.Workflow/Workflow.cs

```diff
@@ -0,0 +1,137 @@
+﻿// ------------------------------------------------------------------------
+// Copyright 2023 The Dapr Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ------------------------------------------------------------------------
+
+namespace Dapr.Workflow
+{
+    using System;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Common interface for workflow implementations.
+    /// </summary>
+    /// <remarks>
+    /// Users should not implement workflows using this interface, directly.
+    /// Instead, <see cref="Workflow{TInput, TOutput}"/> should be used to implement workflows.
+    /// </remarks>
+    public interface IWorkflow
+    {
+        /// <summary>
+        /// Gets the type of the input parameter that this workflow accepts.
+        /// </summary>
+        Type InputType { get; }
+
+        /// <summary>
+        /// Gets the type of the return value that this workflow produces.
+        /// </summary>
+        Type OutputType { get; }
+
+        /// <summary>
+        /// Invokes the workflow with the specified context and input.
+        /// </summary>
+        /// <param name="context">The workflow's context.</param>
+        /// <param name="input">The workflow's input.</param>
+        /// <returns>Returns the workflow output as the result of a <see cref="Task"/>.</returns>
+        Task<object?> RunAsync(WorkflowContext context, object? input);
+    }
+
+    /// <summary>
+    /// Represents the base class for workflows.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    ///  Workflows describe how actions are executed and the order in which actions are executed. Workflows
+    ///  don't call into external services or do complex computation directly. Rather, they delegate these tasks to
+    ///  <em>activities</em>, which perform the actual work.
+    /// </para>
+    /// <para>
+    ///   Workflows can be scheduled using the Dapr client or by other workflows as child-workflows using the
+    ///   <see cref="WorkflowContext.CallChildWorkflowAsync"/> method.
+    /// </para>
+    /// <para>
+    ///   Workflows may be replayed multiple times to rebuild their local state after being reloaded into memory.
+    ///   workflow code must therefore be <em>deterministic</em> to ensure no unexpected side effects from execution
```

Comment:
Actually, workflows don't need to be idempotent because they don't have any external side effects, which essentially guarantees _effectively-once_ execution guarantees. The only restriction therefore is that workflows are deterministic.

Activities don't need to be deterministic, but it is a good practice to make them idempotent since they have at-least-once execution guarantees. I believe we discuss this in the comments for the `WorkflowActivity` class.